### PR TITLE
Impl serde

### PR DESCRIPTION
**Pull request by mfornet**
Enable serde behind a feature flag. This is useful for libraries/binaries that use aurora as a dependency.

Note that this only enable serde for types required by borealis refiners. The PR can be extended for all public types or that can be done in demand, wdyt?

### CODE REVIEW

**Comment by joshuajbouw** [comment]
We had attempted to have serde in but at the time, it didn't make sense. Given today's uses where the Engine is now being used as a library, this makes sense in this way.

File: engine-transactions/Cargo.toml [hunk 1]
```diff
@@ -19,6 +19,8 @@ aurora-engine-precompiles = { path = "../engine-precompiles", default-features =
 evm = { git = "https://github.com/aurora-is-near/sputnikvm.git", rev = "37448b6cacd98b06282cff5a559684505c29bd2b", default-features = false }
 rlp = { version = "0.5.0", default-features = false }
 hex = { version = "0.4", default-features = false, features = ["alloc"] }
+serde = { version = "1", features = ["derive"], optional = true }
 
 [features]
 std = ["aurora-engine-types/std", "aurora-engine-precompiles/std", "evm/std", "rlp/std", "hex/std"]
+impl-serde = ["aurora-engine-types/impl-serde", "serde"]
```
[the remark this answers is not in the document]
Comment (joshuajbouw) [reply engine-transactions/Cargo.toml:26]: Agreed with this.

**Comment by joshuajbouw** [comment]
There should be a check that it isn't being compiled as `no_std` too. That should fix some issues. `serde` won't work with it.

File: engine-transactions/Cargo.toml (hunk 1 above)
[the remark this answers is not in the document]
Comment (mfornet) [reply engine-transactions/Cargo.toml:26]: Doing `serde = ["aurora-engine-types/serde", "serde"]` won't work:

```
error: failed to parse manifest at `/Users/marcelo/Documents/near/aurora-engine/engine-types/Cargo.toml`

Caused by:
  optional dependency `serde` is not included in any feature
  Make sure that `dep:serde` is included in one of features in the [features] table.
```

It was done intentional this way. Notice that `serde` is an [optional dependency](https://doc.rust-lang.org/cargo/reference/features.html#optional-dependencies):

`serde = { version = "1", features = ["derive"], optional = true }`

How it works is that it creates a new feature, that if enabled then this dependency is included. However, to enable this dependency we need to enable as well `primitive-types/impl-serde`. If I try to do it directly like:

`serde = { version = "1", features = ["derive", "primitive-types/impl-serde"], optional = true }`

I get the following error:

```
error: failed to parse manifest at `/Users/marcelo/Documents/near/aurora-engine/engine-types/Cargo.toml`

Caused by:
  feature `primitive-types/impl-serde` in dependency `serde` is not allowed to contain slashes
  If you want to enable features of a transitive dependency, the direct dependency needs to re-export those features from the `[features]` table.
```

And the suggestion is: `the direct dependency needs to re-export those features`. So what I'm doing is create a "direct dependency": `impl-serde` which depends both on `serde` and on `primitive-types/impl-serde` (in the case of `engine-types` crate). 

Inside the code I added the dependency directly on `serde` in the rare case `serde` is enabled but `impl-serde` is not, which should work though it is not particularly useful.

**Comment by mfornet** [comment]
> There should be a check that it isn't being compiled as `no_std` too. That should fix some issues. `serde` won't work with it.

How do you propose to do this? I expect serde won't be included unless either `serde` or `impl-serde` features are set.

File: engine-transactions/Cargo.toml (hunk 1 above)
[the remark this answers is not in the document]
Comment (mfornet) [reply engine-transactions/Cargo.toml:26]: There is also the option of [locally renaming dependencies](https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#renaming-dependencies-in-cargotoml), but that means internally we won't be using `use serde::Serialize` anymore, but instead `use renamed_serde::Serialize`.

Maybe I'm missing something else we could do.

Btw, I thought this was happening as well in `primitve-types` but looking closely [they have a different reason](https://github.com/paritytech/parity-common/blob/master/primitive-types/Cargo.toml#L14) to use `impl-serde`.

[the remark this answers is not in the document]
Comment (birchmd) [reply engine-transactions/Cargo.toml:26]: Ah, ok, thanks for the clarification @mfornet 

Still seems strange to me that we need to introduce a sort of proxy feature here in order to propagate down the dependency chain. But if that's what it takes then that's fine I suppose.

Comment (birchmd) [inline engine-transactions/Cargo.toml:26]: Nit: newline at EOF please

File: engine-types/Cargo.toml [hunk 2]
```diff
@@ -28,3 +29,4 @@ rand = "0.7.3"
 [features]
 default = ["std"]
 std = ["primitive-types/std"]
+impl-serde = ["primitive-types/impl-serde", "serde"]
```
Comment (birchmd) [inline engine-types/Cargo.toml:32]: Nit: newline at EOF please

**Review by birchmd — APPROVED** [review]

**Review by joshuajbouw — APPROVED** [review]

**Comment by joshuajbouw** [comment]
Resolve conflicts please @mfornet

**Pull request merged**